Tracking doc overhaul (#10471)
mlflow/mlflow: docs/source/tracking/backend-stores.rst

.. _backend-stores:

==============
Backend Stores
==============

The backend store is a core component in `MLflow Tracking <../index.html>`_ where MLflow stores metadata for 
:ref:`Runs <runs>` and experiments such as:

* Run ID
* Start & end time
* Parameters
* Metrics
* Code version (only if you launch runs from an :ref:`MLflow Project <projects>`).
* Source file name (only if you launch runs from an :ref:`MLflow Project <projects>`).

> Note that large model artifacts such as model weight files are stored in `artifact store <artifacts-stores.html>`_,  the other core component in MLflow Tracking.

Reviewer: ```suggestion
Note that large model artifacts such as model weight files are stored in `artifact store <artifacts-stores.html>`_,  the other core components are stored in your backend store (MLflow Tracking).
```

Author: oops I meant sth like "artifact store, which is the other component in MLflow tracking", will rephase a bit

.. _where_runs_are_recorded:

Configure Backend Store
=======================
By default, MLflow stores metadata in local files in the ``./mlruns`` directory, but MLflow can store metadata to databases as well.

> You can configure the location by passing **tracking URI** to MLflow, via either of the following methods:

Reviewer: ```suggestion
You can configure the location by passing the desired **tracking URI** to MLflow, via either of the following methods:
```

* Set the ``MLFLOW_TRACKING_URI`` environment variable.
* Call :py:func:`mlflow.set_tracking_uri` in your code.

> * If you are running :ref:`Tracking Server <tracking_server>`, you can set the ``tracking_uri`` option when starting the server, like ``mlflow server --backend-store-uri sqlite:///mydb.sqlite``

Reviewer: ```suggestion
* If you are running a :ref:`Tracking Server <tracking_server>`, you can set the ``tracking_uri`` option when starting the server, like ``mlflow server --backend-store-uri sqlite:///mydb.sqlite``
```

Continue to the next section for the supported format of tracking URLs.
Also visit :ref:`this guidance <tracking_setup>` for how to set up the backend store properly for your workflow.

Supported Store Types
=====================
MLflow supports the following types of tracking URI for backend stores:

> - Local file path (specified as ``file:/my/local/dir``), where data is just directly stored locally.

Reviewer: ```suggestion
- Local file path (specified as ``file:/my/local/dir``), where data is just directly stored locally to a system disk where your code is executing.
```

> - Database encoded as ``<dialect>+<driver>://<username>:<password>@<host>:<port>/<database>``. MLflow supports the dialects ``mysql``, ``mssql``, ``sqlite``, and ``postgresql``. For more details, see `SQLAlchemy database uri <https://docs.sqlalchemy.org/en/latest/core/engines.html#database-urls>`_.

Reviewer: ```suggestion
- A Database, encoded as ``<dialect>+<driver>://<username>:<password>@<host>:<port>/<database>``. MLflow supports the dialects ``mysql``, ``mssql``, ``sqlite``, and ``postgresql``. For more details, see `SQLAlchemy database uri <https://docs.sqlalchemy.org/en/latest/core/engines.html#database-urls>`_.
```

- HTTP server (specified as ``https://my-server:5000``), which is a server hosting an :ref:`MLflow tracking server <tracking_server>`.
- Databricks workspace (specified as ``databricks`` or as ``databricks://<profileName>``, a `Databricks CLI profile <https://github.com/databricks/databricks-cli#installation>`_).
  Refer to Access the MLflow tracking server from outside Databricks `[AWS] <http://docs.databricks.com/applications/mlflow/access-hosted-tracking-server.html>`_
  `[Azure] <http://docs.microsoft.com/azure/databricks/applications/mlflow/access-hosted-tracking-server>`_, or :ref:`the quickstart <quickstart_tracking_server>` to
  easily get started with hosted MLflow on Databricks Community Edition.

> 

Reviewer: AzureML?

Author: Hmm as far as I read the linked doc, there is no mention about AzureML... It describes how to use MLflow outside AD

> .. note::

Reviewer: We should change this to "important". So many people ask this question. Let's emphasize text to make it stand out as well.

Author: Got it! Also worth add it to the FAQ section then.

for permanently removing Run metadata and artifacts for deleted runs.

SQLAlchemy Options
==================
You can inject some `SQLAlchemy connection pooling options <https://docs.sqlalchemy.org/en/latest/core/pooling.html>`_ using environment variables.

+-----------------------------------------+-----------------------------+
| MLflow Environment Variable             | SQLAlchemy QueuePool Option |
+-----------------------------------------+-----------------------------+
| ``MLFLOW_SQLALCHEMYSTORE_POOL_SIZE``    | ``pool_size``               |
+-----------------------------------------+-----------------------------+
| ``MLFLOW_SQLALCHEMYSTORE_POOL_RECYCLE`` | ``pool_recycle``            |
+-----------------------------------------+-----------------------------+
| ``MLFLOW_SQLALCHEMYSTORE_MAX_OVERFLOW`` | ``max_overflow``            |
+-----------------------------------------+-----------------------------+

File Store Performance
======================

MLflow will automatically try to use `LibYAML <https://pyyaml.org/wiki/LibYAML>`_ bindings if they are already installed.

> However if you notice any performance issues when using *file store* backend, it could mean LibYAML is not installed on your system.

Reviewer: ```suggestion
However, if you notice any performance issues when using *file store* backend, it could mean LibYAML is not installed on your system.
```
